### PR TITLE
Query: Don't read default value from databases

### LINKED ISSUE #20633: Query: We should throw when accessing non-nullable property over a null-value entity in the projection
query:

```cs
ctx.Gears.Select(g => g.AssignedCity.Name.Length);
```

assuming assigned city can be null, we should throw "nullable object must have a value" (or ideally something more meaningful).

Currently we return default(int) - it might not be what user expected.
Casting result to nullable will yield correct result (and would get rid of the exception in the future)

```cs
ctx.Gears.Select(g => (int?)g.AssignedCity.Name.Length);
```

### LINKED ISSUE #20959: DateTimeOffset to DateTime Null Conversion
When doing a select query into a class with a Nullable<DateTime> from a database model class property that's a Nullable<DateTimeOffset> it seems the conversion wants to return a DateTime.MinValue. I'd expect all of the below tests to return null.

### Steps to reproduce

``` C#
void Main()
{
	var test1 = Coupons.Select(r => new
		{ r.EndDateTime }
	);

	foreach (var t1 in test1)
	{
		Console.WriteLine("Test1");
		Console.WriteLine(t1.EndDateTime);
		Console.WriteLine();
	}

	var test2 = Coupons.Select(c => new DTO
	{
		DT = c.EndDateTime == null ? (DateTime?)null : c.EndDateTime.Value.DateTime.Date
	});

	foreach (var t2 in test2)
	{
		Console.WriteLine("Test2");
		Console.WriteLine(t2.DT);
		Console.WriteLine();
	}

	var test3 = Coupons.ToList().Select(c => new DTO
	{
		DT = c.EndDateTime == null ? (DateTime?)null : c.EndDateTime.Value.DateTime.Date
	});

	foreach (var t3 in test3)
	{
		Console.WriteLine("Test3");
		Console.WriteLine(t3.DT);
		Console.WriteLine();
	}

}
public class DTO
{
	public DateTime? DT { get; set; }
}

// This is a database model class. There's also nothing in the constructor that modifies this property.
public class Coupons {
	public DateTimeOffset? EndDateTime { get; set; }
}
```

### Console

Test1
null

Test2
1/1/0001 12:00:00 AM

Test3
null

### Further technical details

EF Core version: 3.1.2
Database provider:  Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Standard 2.1
Operating system: Windows 10 64bit
IDE: Visual Studio 2019 16.5.5

### LINKED ISSUE #21002: Subquery materializes zero in place of null
When materializing an aggregate subquery a null value is materialized as zero.

### Steps to reproduce

``` C#
class Table
{
    public int Id { get; set; }
}
class Db : DbContext
{
    public DbSet<Table> Table { get; set; }
}
// config
services.AddDbContext<Db>(opts => opts.UseSqlServer(connStr, sql => sql.UseRelationalNulls()));
// the actual query
var q = from t in db.Table
        select new
        {
            A = t.Id,
            B = db.Table
                .Where(t2 => t2.Id > 10000/*some ridiculously large number that makes this condition always false but prevents any optimization*/)
                .Max(t2 => t2.Id),
        };
var r = await q.ToListAsync();
```
The raw sql would look something like this:
``` SQL
SELECT
	[t0].[Id] AS [A],
	(
		SELECT MAX([t].[Id])
		FROM [Table] AS [t]
		WHERE [t].[Id] > 10000
	) AS [B]
FROM [Table] AS [t0]
```
When executed directly on the db it would yield the following values:
A	B
1	null
2	null

However EFCore materializes the following:
A	B
1	0
2	0

EF6 threw an exception in this case (attempted to materialize null as a non-null value, can't remember the exact wording of the exception). Is this behavior intentional?

When the subquery is forced to be nullable, e.g.
``` C#
var q = from t in db.Table
        select new
        {
            A = t.Id,
            B = db.Table
                .Where(t2 => t2.Id > 10000/*some ridiculously large number that makes this condition always false but prevents any optimization*/)
                .Max(t2 => (int?)t2.Id),
        };
```
then the expected null values are materialized in B.

~~I'm also fiddling with the provider model extending the relational base provider and return null values in the DbDataReader however EFCore still tries to call DbDataReader.GetInt32 for the B column (and fails miserably with an exception) regardless of the nullability of B. I'm not sure what's going on here, this looks like different behavior than SQL Server.~~
Scratch that. It was a bug in my code.

### Further technical details

EF Core version: 3.1.4
Database provider: Microsoft.EntityFrameworkCore.SqlServer)
Target framework: .NET Core 3.1
Operating system: Windows 10 (10.0.18363.836)
IDE: Visual Studio 2019 16.6.0

### PR DESCRIPTION
**Pull request by smitpatel**
In 3.x when reading values from database, we started returning default value of expected type if the value was null in the database. That means if we needed to read int then database value was null then we got 0 back. Which is not correct as we should throw exception for the case.

The fix here is,
- All value read from databases are of nullable type as database can return null.
- Since all the bindings with database are nullable now,
  - When doing member access we add null check and return nullable of return type
  - When doing method call we add null check for instance. All parameters we convert to original types.
  - All other expressions, where they need to match the type to reconstruct the original tree, adds convert nodes.
- For translation of Single/SingleOrDefault (and friends)
  - For scalar subquery
    - We add coalesce with default value for SingleOrDefault
    - We don't add coalesce for Single case and assume there will be non-null value coming out of database which we will try to assign.
  - For non-scalar subquery
    - We add default value in client side in shaper for SingleOrDefault
    - We throw exception Sequence contains no elements for Single case
  - Above rules are only when not returning entity type. When returning entity type, we don't have enough information to differentiate the case. Also above behavior is just side-effect of assigning values to properties which cannot take null values.
- When doing DefaultIfEmpty over a non-reference type, navigation expansion will add client side coalesce so that we can move Selector after DefaultIfEmpty

Resolves #20633
Resolves #20959
Resolves #21002

### CODE REVIEW

**Comment by smitpatel** [comment]
Pending TODOs
- Still working on replicating same for InMemory/Cosmos
- Update or remove tests which were incorrect due to 3.x behavior
- Add exception messages in resources file.

File: src/EFCore.Relational/Query/Internal/RelationalProjectionBindingExpressionVisitor.cs [hunk 1]
```diff
@@ -426,5 +573,26 @@ private void VerifySelectExpression(ProjectionBindingExpression projectionBindin
...
         }
+
+        private static Expression MatchTypes(Expression expression, Type targetType)
+        {
+            if (targetType != expression.Type
+                && targetType.TryGetElementType(typeof(IQueryable<>)) == null)
+            {
+                if (targetType.MakeNullable() != expression.Type)
+                {
+                    throw new InvalidFilterCriteriaException();
```
Comment (maumar) [inline src/EFCore.Relational/Query/Internal/RelationalProjectionBindingExpressionVisitor.cs:584]: thats a weird exception to throw here

Comment (smitpatel) [reply src/EFCore.Relational/Query/Internal/RelationalProjectionBindingExpressionVisitor.cs:584]: I will convert weird exceptions to Check.DebugAssert.

File: src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs [hunk 2]
```diff
@@ -460,7 +461,9 @@ protected override Expression VisitExtension(Expression extensionExpression)
                         }
                         else
                         {
-                            accessor = valueParameter;
+                            accessor = //valueParameter.Type != projectionBindingExpression.Type
```
Comment (maumar) [inline src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs:464]: clean up here and above

File: test/EFCore.Specification.Tests/Query/NorthwindMiscellaneousQueryTestBase.cs [hunk 3]
```diff
@@ -6047,5 +6048,29 @@ public virtual Task AsEnumerable_over_string(bool async)
         }
 
         private static int ClientMethod(int s) => s;
+
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual async Task Non_nullable_property_through_optional_navigation(bool async)
+        {
+            Assert.Equal(
```
Comment (maumar) [inline test/EFCore.Specification.Tests/Query/NorthwindMiscellaneousQueryTestBase.cs:6056]: add helper method into QTB? (just like we have for assert translation failed)

Comment (smitpatel) [reply test/EFCore.Specification.Tests/Query/NorthwindMiscellaneousQueryTestBase.cs:6056]: Very few and messages also differ. 
A better DRY I think would be to have a method like AssertQuery or something which wraps AssertQuery, where it takes the () => AssertQuery and verifies InvalidOperationException (or may be generic on type), and returns message so each individual method can verify the message.

**Pull request merged**